### PR TITLE
replicate legacy setup

### PR DESCRIPTION
**Pull request by budzianowski**
1. Init of mlp from brax/ksim-legacy
2. Bring back brax/ksim-legacy action scaling (stable fp32 training on both 4090 and a100)
3. Lower number of trajectory steps to be closer to the proper rollout setup
4. Explicit dt

### CODE REVIEW

File: examples/default_humanoid/walking.py [hunk 1]
```diff
@@ -64,11 +65,10 @@ class HumanoidWalkingConfig(PPOConfig, MjxEnvConfig):
     robot_model_name: str = xax.field(value="examples/default_humanoid/")
 
     # ML model parameters.
-    actor_hidden_dims: int = xax.field(value=512)
-    actor_num_layers: int = xax.field(value=2)
-    critic_hidden_dims: int = xax.field(value=512)
-    critic_num_layers: int = xax.field(value=4)
-    init_noise_std: float = xax.field(value=1.0)
+    actor_hidden_dims: int = xax.field(value=64)
```
Comment (budzianowski) [inline examples/default_humanoid/walking.py:68]: copying your previous setup @michael-lutz

File: ksim/model/mlp.py [hunk 2]
```diff
@@ -28,8 +37,14 @@ def __call__(self, x: jax.Array) -> jax.Array:
         """
         for _ in range(self.num_hidden_layers):
             x = nn.Dense(
-                features=self.hidden_features, kernel_init=nn.initializers.kaiming_normal()
+                features=self.hidden_features,
+                kernel_init=nn.initializers.lecun_uniform(),
```
Comment (budzianowski) [inline ksim/model/mlp.py:43]: Default brax/playground init

File: examples/default_humanoid/walking.py (hunk 1 above)
Comment (michael-lutz) [reply examples/default_humanoid/walking.py:68]: My old setup was hacky, things were actually set to:
```
policy_hidden_layer_sizes=(64,) * 5,
value_hidden_layer_sizes=(64,) * 5,
```
In `ppo.py`. Tbh I don't think it matters, and your defaults are probably better.

File: examples/default_humanoid/walking.py [hunk 3]
```diff
@@ -152,11 +150,18 @@ def get_init_critic_carry(self) -> None:
...
         HumanoidWalkingConfig(
-            num_envs=2048,
-            num_steps_per_trajectory=600,
-            minibatch_size=1024,
-            num_learning_epochs=10,
-            save_every_n_seconds=60 * 30,
+            num_envs=1024,
+            num_steps_per_trajectory=300,
+            minibatch_size=640,
+            num_learning_epochs=1,
```
Comment (michael-lutz) [inline examples/default_humanoid/walking.py:156]: Does it break if we set this > 1, or was this for speed? Brax setup uses 8 epochs.

File: examples/default_humanoid/walking.py [hunk 4]
```diff
@@ -152,11 +150,18 @@ def get_init_critic_carry(self) -> None:
...
+            save_every_n_seconds=60 * 4,
             only_save_most_recent=False,
+            # ksim-legacy original setup was dt=0.003 and ctrl_dt=0.012 ~ 83.33 hz
+            ctrl_dt=0.01,
+            dt=0.001,
```
Comment (michael-lutz) [inline examples/default_humanoid/walking.py:161]: Does it break if we set to 0.003 or 0.005 (default values for `default_humanoid.xml` and `humanoid.xml` respectively)

File: examples/default_humanoid/walking.py [hunk 5]
```diff
@@ -152,11 +150,18 @@ def get_init_critic_carry(self) -> None:
...
+            solver_type=mujoco.mjtSolver.mjSOL_NEWTON.value,
```
Comment (michael-lutz) [inline examples/default_humanoid/walking.py:162]: I think the brax setup does: `mj_model.opt.solver = mujoco.mjtSolver.mjSOL_CG`
does the solver make a difference for you?

File: ksim/env/mjx/actuators/scaled_torque_actuator.py [hunk 6]
```diff
@@ -35,8 +35,11 @@ def __init__(
 
     def get_ctrl(self, mjx_data: mjx.Data, action: Array) -> Array:
         """Get the control signal from the (position) action vector."""
-        clipped_action = jnp.clip(action, self.input_ranges[:, 0], self.input_ranges[:, 1])
-        ctrl = clipped_action * self.gear_ratios
+        # Copy legacy and brax setup here
+        action_min = self.input_ranges[:, 0]
+        action_max = self.input_ranges[:, 1]
+        # Scale action from [-1,1] to actuator limits
```
Comment (michael-lutz) [inline ksim/env/mjx/actuators/scaled_torque_actuator.py:41]: Action does not necessarily range from [-1, 1], it's just the model output which should be normal (in expectation) after obs normalization.

In our case where action_max and action_min are symmetrical, this resolves to action * (action_max - action_min) * .5. All it does is multiple the output by .4 - do we want to just make it explicit? The current code will have weird edge case if the action_max and action_men range are not symmetrical about 0.

File: ksim/model/mlp.py (hunk 2 above)
Comment (michael-lutz) [reply ksim/model/mlp.py:43]: Not sure if we want this. Should be the same as xavier initialization i.f.f. n_in = n_out. Xavier is more general and I think makes more sense here. Since we use relu, I think he would probably make more sense.

Comment (michael-lutz) [reply ksim/model/mlp.py:43]: I'm also fine with orthogonal initialization, as per 37 implementation details

File: ksim/model/mlp.py [hunk 7]
```diff
@@ -28,8 +37,14 @@ def __call__(self, x: jax.Array) -> jax.Array:
...
             )(x)
-            x = nn.relu(x)
-        x = nn.Dense(features=self.out_features, kernel_init=nn.initializers.xavier_normal())(x)
+            x = self.activation(x)
+
+        x = nn.Dense(
+            features=self.out_features,
```
Comment (michael-lutz) [inline ksim/model/mlp.py:48]: See above

**Review by michael-lutz — APPROVED** [review]

File: ksim/model/mlp.py (hunk 2 above)
Comment (budzianowski) [reply ksim/model/mlp.py:43]: I agree - I just want to keep it the same way for now.

File: examples/default_humanoid/walking.py (hunk 3 above)
Comment (budzianowski) [reply examples/default_humanoid/walking.py:156]: This is actually important to get any sign of sane training since a proper ppo step typically is:
2048envs * 10 steps with minibatch 640 with 10 updates -> 204800 datapoints to train on 
Our current setup with this
1024*300 with 1 update - > 307200 datapoints to train on

File: examples/default_humanoid/walking.py (hunk 4 above)
Comment (budzianowski) [reply examples/default_humanoid/walking.py:161]: I haven't tried yet.

File: examples/default_humanoid/walking.py (hunk 5 above)
Comment (budzianowski) [reply examples/default_humanoid/walking.py:162]: Brax default in humanoid is Newton, your legacy is CG. Both work.
